pkg/resources/schemas/template_test.go

```diff
@@ -0,0 +1,375 @@
+				eventsReady := make(chan struct{})
+
+				// initial schemas
+				factory.EXPECT().Schemas(userInfo).Return(types.EmptyAPISchemas(), nil)
+
+				// return updated schemas for the second request
+				factory.EXPECT().Schemas(userInfo).DoAndReturn(func(_ user.Info) (*types.APISchemas, error) {
+					// signal that initial Schemas were called
+					close(eventsReady)
+					return types.EmptyAPISchemas(), nil
+				})
+
+				return testValues{nil, factory, eventsReady}
+			},
+		},
+		{
+			name: "Schema kind attribute is updated",
+			setup: func(ctrl *gomock.Controller) testValues {
+				factory := schemafake.NewMockFactory(ctrl)
+				eventsReady := make(chan struct{})
+				baseSchemas := types.EmptyAPISchemas()
+				updateSchemas := types.EmptyAPISchemas()
+
+				testSchema := types.APISchema{
+					Schema: &v1schema.Schema{
+						ID:                "pod",
+						PluralName:        "pods",
+						CollectionMethods: []string{"GET"},
+						ResourceMethods:   []string{"GET"},
+					},
+				}
+				baseSchemas.AddSchema(testSchema)
+				// initial schemas
+				factory.EXPECT().Schemas(userInfo).Return(baseSchemas, nil)
+
+				// add kind attribute
+				attributes.SetKind(&testSchema, "newKind")
+				updateSchemas.AddSchema(testSchema)
+
+				// return updated schemas for the second request
+				factory.EXPECT().Schemas(userInfo).DoAndReturn(func(_ user.Info) (*types.APISchemas, error) {
+					// signal that initial Schemas were called
+					close(eventsReady)
+					return updateSchemas, nil
+				})
+
+				expectedEvents := []types.APIEvent{
+					{
+						Name:         types.ChangeAPIEvent,
+						ResourceType: "schema",
+						Object: types.APIObject{
+							Type:   resourceType,
+							ID:     testSchema.ID,
+							Object: &testSchema,
+						},
+					},
+				}
+				return testValues{expectedEvents, factory, eventsReady}
+			},
+		},
+		{
+			name: "Schema access attribute is updated",
+			setup: func(ctrl *gomock.Controller) testValues {
+				factory := schemafake.NewMockFactory(ctrl)
+				eventsReady := make(chan struct{})
+				baseSchemas := types.EmptyAPISchemas()
+				updateSchemas := types.EmptyAPISchemas()
+
+				testSchema := types.APISchema{
+					Schema: &v1schema.Schema{
+						ID:                "pod",
+						PluralName:        "pods",
+						CollectionMethods: []string{"GET"},
+						ResourceMethods:   []string{"GET"},
+					},
+				}
+				baseSchemas.AddSchema(testSchema)
+				// initial schemas
+				factory.EXPECT().Schemas(userInfo).Return(baseSchemas, nil)
+
+				// add access attribute
+				attributes.SetAccess(&testSchema, map[string]string{"List": "*"})
+				updateSchemas.AddSchema(testSchema)
+
+				// return updated schemas for the second request
+				factory.EXPECT().Schemas(userInfo).DoAndReturn(func(_ user.Info) (*types.APISchemas, error) {
+					// signal that schemas were requested
+					close(eventsReady)
+					return updateSchemas, nil
+				})
+
+				expectedEvents := []types.APIEvent{}
```

Comment:
I'm surprised by this, can you help me understand why this is expected behavior? `schemasForSubject` is all about regenerating the schemas if the user's access changed. If the user lost access to an API type, I would expect the schemas to need a refresh.

Reply:
This was the cause of the original bug. When we create the schemas with `schema.Factory.Schemas(user)` we add the specific verbs a user has on each schema in the attributes section of the schema under the key `access` [here](https://github.com/rancher/steve/blob/06584d207dac89091e5601884282bf7e109e19f0/pkg/schema/factory.go#L134).
When returning each schema as an APIEvent over the websocket we must first convert the Schema into an APIObject in apiserver [here](https://github.com/rancher/apiserver/blob/fe12f2183e7112becec1633eb2ea728a09cece47/pkg/store/schema/schema_store.go#L19). For reason I am not positive on we deliberately strip off the `access` key from the attributes during this conversion. Thus when Steve was comparing the APIObject schema and the previously recorded Schema one would be missing the `access` verbs and a change would always be detected.

I ended up stripping the `access` key out of attributes on the old Schema as well before the comparison so we could get accurate change detection. I strayed away from removing the delete during conversion in apiserver for 2 reasons.
1. I am not sure if Steve is the only consumer of `apiserver.APISchemas`
2. The events sent to the UI previously did not include the `access` verbs in the attributes, so I didn't want to add new data that may not be used.

Reply:
Also, if the user's access changes, we regenerate the available schemas for the user via `userChangeNotify` in `template.go`. The above test just demonstrates that we do not send the verbs that changed for each schema.

Reply:
Spent some time playing with this. It looks like if I add or remove a role for a user, the `resource.create` and `resource.remove` events go through (because there's nothing to call DeepEqual on). If I change the role's verbs from, e.g. [create, update, get, list] to just [get, list], the `resource.change` event also goes through because the object's ResourceMethods and CollectionMethods change and the DeepEqual returns false. On the other hand, if I change it from [get, list] to just [get], the methods do not change, so no event goes through. But this seems fine, since as you pointed out the UI isn't using the access verbs anyways. The schema change event doesn't affect the user's actual access level. So I think this makes sense :+1: